tap/passive_tapper.go

```diff
@@ -198,8 +201,59 @@ func startMemoryProfiler() {
 	}()
 }
 
+func closeTimedoutTcpStreamChannels() {
+	cpuCount := runtime.NumCPU()
+	// +1 Stats tracker Goroutine
+	// +1 Main packet capture loop
+	baseGoroutineCount := runtime.NumGoroutine() + 2
+	for {
+		time.Sleep(1 * time.Millisecond)
+		streams.Range(func(key interface{}, value interface{}) bool {
+			streamWrapper := value.(*tcpStreamWrapper)
+			stream := streamWrapper.stream
+			// `diff` tries to stabilize towards 0.
+			// `diff` means the number of currently running extra Goroutines mainly spawned for dissectors.
+			n := runtime.NumGoroutine()
+			diff := n - baseGoroutineCount
+			if diff < 1 {
+				diff = 1
+			}
+			// `metric` tries to stabilize arround 1.0.
+			// `metric` means what percentage of the load can be actually handled by the given resources.
+			metric := float64(cpuCount) / float64(diff)
+			dynamicStreamChannelTimeoutMs := int(float64(baseStreamChannelTimeoutMs) * metric)
+			dynamicStreamChannelTimeoutNs := dynamicStreamChannelTimeoutMs * 1000000
```

Comment:
`dynamicStreamChannelTimeoutNs` is difficult to reason about.
A constant timeout makes more sense to me. We can say that if a protocol was not detected in under X ms and/or after Y packets, then the stream should be dropped.
I expect each routine to be blocked most of the time, waiting for traffic, so a time-based criterion is more suitable than guessing the effective CPU cycles dedicated to dissector.

Reply:
@nimrod-up9 replaced the formula with hard timeout 5000 milliseconds and 4000 maximum number of Goroutines https://github.com/up9inc/mizu/commit/e92410688c651d34fc53f162f31aa5277e4cdb72